central/complianceoperator/v2/report/manager/manager_impl.go

```diff
@@ -198,7 +207,50 @@ func (m *managerImpl) runReports() {
 				continue
 			}
 			logging.Infof("Executing report %q at %v", req.scanConfig.GetId(), time.Now().Format(time.RFC822))
-			go m.generateReport(req)
+			var w watcher.ScanConfigWatcher
+			concurrency.WithLock(&m.watchingScanConfigsLock, func() {
+				w = m.watchingScanConfigs[req.scanConfig.GetId()]
+			})
+			// These will be modified in a follow-up PR
+			reportType := storage.ComplianceOperatorReportStatus_ON_DEMAND
+			notificationMethod := storage.ComplianceOperatorReportStatus_EMAIL
+			snapshot := &storage.ComplianceOperatorReportSnapshotV2{
+				ReportId:            uuid.NewV4().String(),
+				ScanConfigurationId: req.scanConfig.GetId(),
+				Name:                req.scanConfig.GetScanConfigName(),
+				Description:         req.scanConfig.GetDescription(),
+				ReportStatus: &storage.ComplianceOperatorReportStatus{
+					RunState:                 storage.ComplianceOperatorReportStatus_WAITING,
+					StartedAt:                protocompat.TimestampNow(),
+					ReportRequestType:        reportType,
+					ReportNotificationMethod: notificationMethod,
+				},
+				User: req.scanConfig.GetModifiedBy(),
+			}
+			if w == nil {
+				snapshot.GetReportStatus().RunState = storage.ComplianceOperatorReportStatus_PREPARING
+				if err := m.snapshotDataStore.UpsertSnapshot(req.ctx, snapshot); err != nil {
+					log.Errorf("Unable to upsert snapshot: %v", err)
+					continue
```

Comment:
🛑  :bug: I think we must release the semaphore on each continue. Currently the only place where we release them is inside `generateReport`, but if we hit a continue case, we never release.

This even stronger suggests that this should be exported into separate function(s), so that we can use deferred release.

Reply:
Please maybe test this with `ROX_REPORT_EXECUTION_MAX_CONCURRENCY` set to 1.

Reply:
I've refactored to release the semaphore with a defer statement.